clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -2412,6 +2412,150 @@ bool SemaHLSL::CheckCompatibleParameterABI(FunctionDecl *New,
   return HadError;
 }
 
+// Follows PerformScalarCast
+bool SemaHLSL::CanPerformScalarCast(QualType SrcTy, QualType DestTy) {
+
+  if (SemaRef.getASTContext().hasSameUnqualifiedType(SrcTy, DestTy))
+    return true;
+
+  switch (Type::ScalarTypeKind SrcKind = SrcTy->getScalarTypeKind()) {
+  case Type::STK_MemberPointer:
+    return false;
+
+  case Type::STK_CPointer:
+  case Type::STK_BlockPointer:
+  case Type::STK_ObjCObjectPointer:
```

Comment:
HLSL definitely doesn't have block or objective-c object pointers. I also wouldn't expect C pointers to show up in HLSL because we don't decay arrays.

In general I would expect these casts to not operate on pointers of any kind. By the time we add pointers to the language hopefully these casts will be simplified/removed.